ostd/src/arch/x86/pci.rs

```diff
@@ -1,8 +1,36 @@
 // SPDX-License-Identifier: MPL-2.0
 
-//! PCI bus io port
+//! PCI bus access
 
 use super::device::io_port::{IoPort, ReadWriteAccess, WriteOnlyAccess};
+use crate::{bus::pci::PciDeviceLocation, prelude::*};
 
-pub static PCI_ADDRESS_PORT: IoPort<u32, WriteOnlyAccess> = unsafe { IoPort::new(0x0CF8) };
-pub static PCI_DATA_PORT: IoPort<u32, ReadWriteAccess> = unsafe { IoPort::new(0x0CFC) };
+static PCI_ADDRESS_PORT: IoPort<u32, WriteOnlyAccess> = unsafe { IoPort::new(0x0CF8) };
+static PCI_DATA_PORT: IoPort<u32, ReadWriteAccess> = unsafe { IoPort::new(0x0CFC) };
+
+const BIT32_ALIGN_MASK: u32 = 0xFFFC;
+
+pub fn write32(location: &PciDeviceLocation, offset: u32, value: u32) -> Result<()> {
+    PCI_ADDRESS_PORT.write(encode_as_port(location) | (offset & BIT32_ALIGN_MASK));
+    PCI_DATA_PORT.write(value.to_le());
+    Ok(())
+}
+
+pub fn read32(location: &PciDeviceLocation, offset: u32) -> Result<u32> {
+    PCI_ADDRESS_PORT.write(encode_as_port(location) | (offset & BIT32_ALIGN_MASK));
+    Ok(PCI_DATA_PORT.read().to_le())
+}
+
+/// By encoding bus, device, and function into u32, user can access a PCI device in x86 by passing in this value to the address port.
+#[inline(always)]
```

Comment:
According to the [Rust standard library policy](https://std-dev-guide.rust-lang.org/policy/inline.html) or this [well-written blog](https://matklad.github.io/2021/07/09/inline-in-rust.html), `#[inline]` is only useful on public, small, non-generic methods, and is generally not useful on private methods (since the compiler will try to make smart choices automatically).

So I think we should just avoid using `#[inline]` here.

Reply:
I have repeated posted the same argument. This is a common pitfall. It should be written in our Contributing page so that we don't have to repeat ourselves.